openedx/core/lib/block_structure/transformer.py

```diff
@@ -125,4 +126,55 @@ def transform(self, usage_info, block_structure):
                 block structure, with already collected data for the
                 transformer, that is to be transformed in place.
         """
-        pass
+        raise NotImplementedError
+
+
+class OptimizedTransformer(BlockStructureTransformer):
+    """
+    Transformers may optionally choose to subclass this class instead of the
+    base, if their transform logic can be broken apart into a lambda for
+    optimization of combined tree traversals.
+
+    For performance reasons, developers should try to subclass this
+    class instead of the above base class, whenever possible - since
+    with this alternative, traversal of the entire block structure happens
+    only once for all transformers that implement this form of transform.
+    """
+
+    def transform(self, usage_info, block_structure):
```

Comment:
My thinking was that these new transformers *can* be used in the new way (combine filter lambdas, then traverse all at once), but they don't *have* to be used that way, in case other consumers do not use this new paradigm.

Is `transformers.py` the only place we ever plan on transforming? Can we assume that all consumers of `BlockStructureTransformer` will know about and be able to handle this new subclass?